packages/visx-xychart/src/components/series/BarStack.tsx

```diff
@@ -0,0 +1,209 @@
+      ) as [number, number],
+    [combinedData],
+  );
+
+  // stack data
+  const stackedData = useMemo(() => {
+    const hasSomeNegativeValues =
+      comprehensiveDomain.length > 0 && comprehensiveDomain.some(num => num < 0);
+
+    const stack = d3stack<CombinedStackData<XScale, YScale>, string>();
+    stack.keys(dataKeys);
+    if (order) stack.order(stackOrder(order));
+    if (offset || hasSomeNegativeValues) stack.offset(stackOffset(offset || 'diverging'));
+
+    return stack(combinedData);
+  }, [combinedData, dataKeys, comprehensiveDomain, order, offset]);
+
+  // register all child data using the stack-transformed values
+  useEffect(() => {
+    const dataToRegister = getBarStackRegistryData(stackedData, comprehensiveDomain, horizontal);
+    registerData(dataToRegister);
+
+    // unregister data on unmount
+    return () => unregisterData(Object.keys(dataToRegister));
+  }, [
+    comprehensiveDomain,
+    horizontal,
+    stackedData,
+    registerData,
+    unregisterData,
+    barSeriesChildren,
+  ]);
+
+  // register mouse listeners
+  const { showTooltip, hideTooltip } = useContext(TooltipContext) ?? {};
+
+  const handleMouseMove = useCallback(
+    (params: HandlerParams | undefined) => {
+      const { svgPoint } = params || {};
+
+      // invoke showTooltip for each key so all data is available in context,
+      // and let Tooltip find the nearest point among them
+      dataKeys.forEach(key => {
+        const entry = dataRegistry.get(key);
+        const childData = barSeriesChildren.find(child => child.props.dataKey === key)?.props.data;
+        if (childData && svgPoint && width && height && showTooltip) {
+          const datum = findNearestStackDatum(
+            {
+              point: svgPoint,
+              data: entry.data,
+              xScale,
+              yScale,
+              xAccessor: entry.xAccessor,
+              yAccessor: entry.yAccessor,
+              width,
+              height,
+            },
+            childData,
+            horizontal,
+          );
+
+          if (datum) {
+            showTooltip({
+              key,
+              svgPoint,
+              ...datum,
+            });
+          }
+        }
+      });
+    },
+    [
+      barSeriesChildren,
+      dataRegistry,
+      dataKeys,
+      horizontal,
+      xScale,
+      yScale,
+      width,
+      height,
+      showTooltip,
+    ],
+  );
+  useEventEmitter('mousemove', handleMouseMove);
+  useEventEmitter('mouseout', hideTooltip);
+
+  // if scales and data are not available in the registry, bail
+  if (dataKeys.some(key => dataRegistry.get(key) == null) || !xScale || !yScale || !colorScale) {
+    return null;
+  }
+
+  const barThickness = getBandwidth(horizontal ? yScale : xScale);
+
+  return (
+    <g className="visx-bar-stack">
+      {stackedData.map((barStack, stackIndex) => {
+        const entry = dataRegistry.get(barStack.key);
+
+        return entry
+          ? barStack.map((bar, index) => {
```

Comment:
this mirrors the logic in `BarStack/BarStackHorizontal`

Reply:
Perhaps do `horizontal` and other conditions outside the loop when possible.

```ts
const thickness = getBandwidth(horizontal ? yScale : xScale);
const halfThickness = thickness / 2;

let widthFn;
let heightFn;
let xFn;
let yFn;

if (horizontal) {
  widthFn = bar => (xScale(getSecondItem(bar)) || 0) - (xScale(getFirstItem(bar)) || 0);
  heightFn = () => thickness;
  xFn = bar => xScale(getFirstItem(bar));
  yFn = 'bandwidth' in yScale 
    ? bar => yScale(getStackValue(bar.data))
    : bar => Math.max((yScale(getStackValue(bar.data)) || 0) - halfThickness);
} else {
  widthFn = () => thickness;
  heightFn = bar => (yScale(getFirstItem(bar)) || 0) - (yScale(getSecondItem(bar)) || 0);
  xFn = 'bandwidth' in xScale
    ? bar => xScale(getStackValue(bar.data))
    : bar => Math.max((xScale(getStackValue(bar.data)) || 0) - halfThickness);
} 

return (
  <g className="visx-bar-stack">
    {stackedData.map((barStack, stackIndex) => {
      const entry = dataRegistry.get(barStack.key);
      if (!entry) return null;

      const y = yFn ?? (bar => yScale(entry.yAccessor(bar)));

      return barStack.map((bar, index) => {
        const barX = xFn(bar);
        if (!isValidNumber(barX)) return null;
        const barY = y(bar);
        if (!isValidNumber(barY)) return null;

        return (
          <rect
            key={`${stackIndex}-${barStack.key}-${index}`}
            x={barX ?? 0}
            y={barY ?? 0}
            width={widthFn(bar)}
            height={heightFn(bar)}
            fill={colorScale(barStack.key)}
            stroke="transparent"
          />
        );        
      });
    })}
  </g>
);
```

Reply:
great suggestion, also the special `yFn` handling made me realize I didn't actually need the `entry` accessor function 👍